### PR TITLE
605 ipynb pep8bear

### PR DESCRIPTION
**Pull request by caichinger**
Hi,

this PR adresses #605 (PEP8Bear for Jupyter Notebooks).

Some thoughts/questions regarding the current implementation:
- I use https://github.com/jupyter/nbformat for to read/write the notebook, thereby we avoid problems with different notebook versions. However, nbformat is then a necessary requirement, is this okay?
- Since the notebook related part is independent of the bear, I think it should be located somewhere else. However, I do not know where. :)
- There is currently some code duplication related to the original PEP8Bear. At the sprint at the EuroPython @sils1297 said that this approach is easier for now. I do not know whether the bears-using-bears idea is currently being worked on.

Thank you for your help!

### CODE REVIEW

File: tests/python/PEP8JupyterNotebookBearTest.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.5.2"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 0
+}
+
+"""
```
Comment (sils1297) [inline tests/python/PEP8JupyterNotebookBearTest.py]: two newlines here?

File: bears/python/PEP8JupyterNotebookBear.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        Writes notebook in dict form to string array.
+
+        :param file_dct: The notebook in dictionary form.
+        :return:         The notebook in string array form.
+        """
+        file_lst = (nbformat.writes(file_dct, nbformat.NO_CONVERT)
+                            .splitlines(True))
+        # nbformat.writes does not add a final \n, we do this manually
+        # necessary for the diff because original file has it
+        if self.file[-1].endswith('\n') and not file_lst[-1].endswith('\n'):
```
Comment (sils1297) [inline bears/python/PEP8JupyterNotebookBear.py]: youu could add a testcase without a trailing newline at all to cover this branch

Comment (caichinger) [inline bears/python/PEP8JupyterNotebookBear.py]: I will do that

File: tests/python/PEP8JupyterNotebookBearTest.py (hunk 1 above)
Comment (caichinger) [inline tests/python/PEP8JupyterNotebookBearTest.py]: I thought it would simplify things but it actually doesn't, will be changed

**Comment by sils1297** [comment]
ouh this isn't WIP, you changed the title after pushing which is why this was unnoticed for so long - sorry

**Comment by sils1297** [comment]
@caichinger are you still around? Otherwise I'd take this over as we would love it for the release :)

File: bears/python/PEP8NotebookBear.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class PEP8NotebookBear(LocalBear):
+    LANGUAGES = {"Python", "Python 2", "Python 3"}
+    REQUIREMENTS = {PipRequirement('autopep8', '1.*'),
+                    PipRequirement('nbformat', '4.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_FIX = {'Formatting'}
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: I think we'll want an ASCIINEMA_URL here :)

File: bears/python/PEP8NotebookBear.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    def run(self, filename, file,
+            max_line_length: int=79,
+            tab_width: int=SpacingHelper.DEFAULT_TAB_WIDTH,
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: this has been renamed `indent_size`

Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: CC @Asnelchristian

**Comment by caichinger** [comment]
@sils1297 yes I am still around and will incorporate your comments during the weekend (most likely tomorrow) -  next release sounds great! :)
ad "changing title after pushing": I thought changing title alone is sufficient, next time I know.

File: bears/python/PEP8NotebookBear.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        Reads a notebook from a string array and returns the NotebookNode
+        object.
+
+        :param file: The notebook file contents as JSON string array.
+        :return:     The notebook as NotebookNode.
+        """
+        return nbformat.reads(''.join(file), nbformat.NO_CONVERT)
+
+    def write_to_list(self, file_nb):
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: shouldn't this function as a member function simply use self.file_nb instead of receiving it as a param?

Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: why does this receive a parameter that is retrieved from the class? Seems super weird to me

File: bears/python/PEP8NotebookBear.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        file_lst = (nbformat.writes(file_nb, nbformat.NO_CONVERT)
+                            .splitlines(True))
+        # nbformat.writes does not add a final \n, we do this manually
+        # necessary for the diff because original file has it
+        if self.file[-1].endswith('\n') and not file_lst[-1].endswith('\n'):
+            file_lst[-1] += '\n'
+        return file_lst
+
+    @property
+    def file_nb(self):
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: it isn't particularly obvious what this is

File: bears/python/PEP8NotebookBear.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+        :param filename: The filename of the file.
+        :param file:     The file contents as string array.
+        """
+        self.filename = filename
+        self.file = file
+        self._file_nb = self.read_from_list(file)
+
+    @staticmethod
+    def read_from_list(file):
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: is this meant to be public?

**Comment by AbdealiJK** [comment]
I think that the Jupyter notebook idea is awesome !
Would it not e cooler to make a geeric NoteBookCellBear which gives HiddenResults of cells and then we can do a JupyterMarkdownBear, JupyterPythonBear, etc. ? Which takes these values and feeds it into MarkdownBear, PEP8Bear, etc. for analysis.

@sils1297 Any idea on how to do this cleanly ?

Te reason for this is because jupyter notebooks can benefit from all the bears related to R, Julia, Python, Markdown, and Natural Language !

**Comment by sils1297** [comment]
@AbdealiJK we've thought a long time about this so far. Long story short: coala doesn't support this yet. I think @mr-karan wanted to look into this long term. It's complicated - but indeed this bear is a very good thing to drive this concept forward: a nice example. We now know way better about the problems that come with that, e.g. you'd probably need to translate the locations of the results somehow or the bear would have to know about "regions" that are to be analyzed. Great things ahead.

**Comment by AbdealiJK** [comment]
Nod, agree. Is there an issue somewhere for the long term plan ?

**Comment by sils1297** [comment]
https://github.com/coala-analyzer/coala/issues/2538

File: bears/python/PEP8NotebookBear.py (hunk 5 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: I am not entirely sure what you mean by *a parameter that is retrieved from the class* but I had two things in mind:

1. the Bear works with a copy of `_file_nb` and when he has done his job, `file_nb` has changed, i.e. 
`file_nb != NotebookHandler(filename, file).file_nb`.
1. due to the issue with the final newline charater '\n' (cf. line 50) the `write_to_list` method needs to know how the original `file` looked like.

since I need both the new `file_nb` as well as the original `file` to properly construct the needed list (string array) I thought this is simplest solution. does this answer your question?

in any case, I have no objections against changing the approach :)
maybe rename the parameter to something like `file_nb_changed` to avoid confusion?

File: bears/python/PEP8NotebookBear.py (hunk 6 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: it returns the file in notebook-node-form and I used `@property` to make sure that one only receives a copy to work with and does not change the original version of the 
notebook, i.e. `_file_nb` which is created at initialization.

what is you opinion?

apart from that: do you think this is redundant or maybe `file` be treated the same (to be consistent)?

File: bears/python/PEP8NotebookBear.py (hunk 7 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: since nobody else uses it probably not ;)
I will change it to `_read_from_list`.

File: bears/python/PEP8NotebookBear.py (hunk 3 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: of course we do :)
I will look into this

**Comment by caichinger** [comment]
@sils1297 : just in case you overlooked the changes, I am still around (and 
willing) to work on this. However, if you prefer to take over to speed things
up feel free to do so. Otherwise I am happy to receive your feedback. :)


When this Bear is merged I would open a new issue as shown below. I think it is a 
necessary (and feasible) intermediate step to arrive at more general applications like the 
ones already mentioned in https://github.com/coala-analyzer/coala/issues/2538 and @AbdealiJK comment above. 
Are you both fine with that?


Prospective issue:
> Refactor notebook handling from PEP8NotebookBear into separate module

> To pave the way for future coala features dealing with Jupyter Notebooks,
some notebook-specific io functionality has to be provided.

> As the first bear dedicated to notebooks, the `PEP8NotebookBear` module currently 
holds the `NotebookHandler` class. However, since the Jupyter Notebook is language 
agnostic (note that Jupyter stands for Julia, Python and R) it makes sense to 
refactor the notebook handling into a separate module.

**Comment by sils1297** [comment]
I'm kind of around but I'm today busy with a client and tomorrow as well
maybe somebody else can help you? CC @AbdealiJK

**Comment by caichinger** [comment]
No problem! I am not in a hurry. :) If one of you posts some comments I will try to deal with them during the weekend.

**Comment by sils1297** [comment]
For now I'd keep it in the same module

File: bears/python/PEP8NotebookBear.py (hunk 5 above)
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: so at least if I look at this initially I do get the feeling of "complication" but I can't objectively reason about this yet because I don't fully understand what exactly happens where. Usually if my intuition goes like that there is at least a minor simplification that helps understanding a lot.

I'll have to look at it later with more time :/

File: bears/python/PEP8NotebookBear.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        self.file = file
+        self._file_nb = self._read_from_list(file)
+
+    @staticmethod
+    def _read_from_list(file):
+        """
+        Reads a notebook from a string array and returns the NotebookNode
+        object.
+
+        :param file: The notebook file contents as JSON string array.
```
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: A json string array is confusing IMO. "Lines of the notebook file contents which uses JSON format" maybe ?
(I thought `file` was a JSON string array ... hence the confusion)

File: bears/python/PEP8NotebookBear.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+class NotebookHandler:
+    """Utility class for Jupyter Notebook handling."""
+
+    def __init__(self, filename, file):
+        """
+        Constructs a notebook handler to deal with NotebookNode objects.
+        A NotebookNode is a dict-like node with attribute-access.
+
+        :param filename: The filename of the file.
```
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: Is the filename and file needed ? It seems like filename is not being used in the NotebookHandler.

File: bears/python/PEP8NotebookBear.py (hunk 6 above)
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: A docstring here is useful in such cases.
Possibly just `notebook_node` would be a fine name too, but I have nothing against `file_nb`.

The only reason I have to prefer notebook_node is that it  doesnt confuse with the `file` which is a list of lines

File: bears/python/PEP8NotebookBear.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        :param file:     The file contents as string array.
+        """
+        self.filename = filename
+        self.file = file
+        self._file_nb = self._read_from_list(file)
+
+    @staticmethod
+    def _read_from_list(file):
```
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: Better name would be appreciated here !
What about just `_create_notebook_node` as it literally creates it ? Or maybe `_parse_notebook` or `_parse_file` ? Because nbformat does parse the JSON file and gives a nice API for it

File: bears/python/PEP8NotebookBear.py (hunk 5 above)
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: I can see how this makes sense.
I'd recommend changing this function to `_format_notebook` so that it complements the earlier `_parse_notebook`.
Other alternatives: `_format_file()`, `_create_notebook_file` ?

Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: I think to avoid confusion, it would be a good idea to add `newline__at_eof` or so. If it is True/False, use that.
If it is `None`, use `self.file[-1].endswith('\n')`

Also, a more descriptive docstring mentioning it auto-understands the EOF \n

File: bears/python/PEP8NotebookBear.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            indent_size: int=SpacingHelper.DEFAULT_TAB_WIDTH,
+            pep_ignore: typed_list(str)=(),
+            pep_select: typed_list(str)=(),
+            local_pep8_config: bool=False):
+        """
+        Detects and fixes PEP8 incompliant code in Jupyter Notebooks. This bear
+        will not change functionality of the code in any way.
+
+        :param max_line_length:   Maximum number of characters for a line.
+        :param indent_size:         Number of spaces per indent level.
```
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: 2 extra spaces before "Number"

File: bears/python/PEP8NotebookBear.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        diffs = Diff.from_string_arrays(file, corrected).split_diff()
+
+        for diff in diffs:
+            yield Result(self,
+                         "The code does not comply to PEP8.",
+                         affected_code=(diff.range(filename),),
+                         diffs={filename: diff})
+
+    @staticmethod
+    def _bypass_no_newline_at_end_of_file(f):
```
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: This is unusual. Why a decorator ? It may make sense to have a contextmanager here ?
Or just put those 2-3 lines in the code itself ... as it is being used only once.

File: bears/python/PEP8NotebookBear.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        """
+        Utility function to first append and finally remove a '\n' to
+        a given source string in order to avoid 'no newline at end of file'
+        complaint which does not make sense for notebook cells.
+        """
+        def handle_line_breaks(source):
+            source += '\n'
```
Comment (AbdealiJK) [inline bears/python/PEP8NotebookBear.py]: Don't you want to check if there is already a \n ?

File: tests/python/PEP8NotebookBearTest.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+def prepare_string(file):
+    """
+    Utility function to prepare Notebook strings for further processing.
+
+    replace() necessary to avoid
+    nbformat.reader.NotJSONError: Notebook does not appear to be JSON
+    which is caused by
+    json.decoder.JSONDecodeError: Invalid control character
+    which is caused by the \n characters inside the input cells
```
Comment (AbdealiJK) [inline tests/python/PEP8NotebookBearTest.py]: This is interesting. Does this not happen when you read the file directly ? Shouldnt this be in the bear ?

**Comment by caichinger** [comment]
@sils1297 @AbdealiJK thank you for your comprehensive comments!
I already have some ideas and should be able to push tomorrow (please wait until have updated :)).

File: tests/python/PEP8NotebookBearTest.py (hunk 14 above)
Comment (caichinger) [inline tests/python/PEP8NotebookBearTest.py]: This is not only interesting, it is also annoying. ;)
It does not happen when you read the file directly (i.e. when you pass a file directly to the bear).
If you know a better solution I am happy to change this.

**Comment by caichinger** [comment]
Hi there,

after considering your valuable feedback (thanks again!) I decided to rework the module a bit.

What changed and why:

- I dropped the NotebookHandler class and went for two simple functions for 
reading (`read_notebook_node_from_file`) and writing 
(`write_notebook_node_to_file`) the notebook.
Both are named read*/write* because the underlying nbformat functions are named
almost the same (for me this is more natural than calling them parse, or format
but if you disagree I will of course rename them):
	- http://nbformat.readthedocs.io/en/latest/api.html#nbformat.reads
	- http://nbformat.readthedocs.io/en/latest/api.html#nbformat.writes

Both are named *_from_file/*_to_file because within a Bear a file is understood
as a list containing the content of a file linewise.

- The issue with final newline characters within a notebook cell is now dealt
with directly in the `autopep8_fix_code`(the decorator is gone) (thanks for pointing out the case that there also could be too many). See also the 
corresponding comment within the function. There might be a more elegant 
solution but I do not know one.

- The issue related to the final newline character in the corrected file 
(not cell) is handled just before creating the diff (which led me to removing
the NotebookHandler).

- The `run()` method is now a bit more verbose but I am not sure if moving there 
is really a benefit in moving parts aways.

To me current solution is cleaner (and simpler). Please let me know what you 
think about it. :)

File: bears/python/PEP8NotebookBear.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    Reads a notebook from a string array and returns the NotebookNode
+    object.
+
+    :param file: The notebook file contents as string array.
+    :return:     The notebook as NotebookNode.
+    """
+    return nbformat.reads(''.join(file), nbformat.NO_CONVERT)
+
+
+def write_notebook_node_to_file(notebook_node):
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: it doesn't write to file

**Comment by sils1297** [comment]
@caichinger yes indeed - way better!

File: bears/python/PEP8NotebookBear.py (hunk 15 above)
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: i.e. I think you'll want to rename it (sorry for being so brief, hope it's ok)

File: bears/python/PEP8NotebookBear.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+            # autopep8.fix_code automatically adds a final newline at the end,
+            # e.g. `autopep8.fix_code('a=1')` yields 'a = 1\n'.
+            # This is not related to the 'W292' flag, i.e.
+            # `autopep8.fix_code('a=1', options=dict(ignore=('W292',)))` gives
+            # the same result.
+            # For notebook cells, this behaviour does not make sense, hence
+            # newline is removed if `source` does not end with one.
+            if not source.endswith('\n'):
+                return source_corrected[:-1]
+            return source_corrected
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: hm might it be cleaner to have this outside the class and pass the variables explicitly rather than in the closure? https://www.python.org/dev/peps/pep-0020/

File: bears/python/PEP8NotebookBear.py (hunk 15 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: Well it writes what is called a file within a Bear, namely a list containing the content of the code-text-file linewise. The *write* is motivated by the use of http://nbformat.readthedocs.io/en/latest/api.html#nbformat.writes.
I wanted to stay with the names used in the Bears and `nbformat`.

However, if this remains a source of confusion we rename to e.g.:
- read -> parse?
- write -> format?
- file -> string_array?

@AbdealiJK : you would prefer it like this?

Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: @sils1297 : no need to worry about brevity! :)
okay, I will await your opinions on the new names and then rename both functions.

Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: notebook_node_to_string or to_list or so?

Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: I will give `notebook_node_to_list` and `notebook_node_from_list` a try. Should not sound fancier than it is...

File: bears/python/PEP8NotebookBear.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+# dictionaries, but allow attribute access. The structure of these objects
+# matches the notebook format specification.
+
+
+def notebook_node_from_list(file):
+    """
+    Reads a notebook from a string array and returns the NotebookNode
+    object.
+
+    :param file: The notebook file contents as string array.
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: rename the parameter to string_list maybe?

**Comment by sils1297** [comment]
oh btw. circleci fails because there's some issue with one of the doc comments, see https://circleci.com/gh/coala-analyzer/coala-bears/2824 for more info, it's apparently not valid restructured text

File: bears/python/PEP8NotebookBear.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+                                     .splitlines(True))
+    return notebook_node_as_string_array
+
+
+def autopep8_fix_code_cell(source, options=None, apply_config=None):
+    """
+    Applies autopep8.fix_code and takes care of newline characters.
+
+    autopep8.fix_code automatically adds a final newline at the end,
+    e.g. `autopep8.fix_code('a=1')` yields 'a = 1\n'.
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: use double backticks as we're in rst here

File: bears/python/PEP8NotebookBear.py [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+
+
+class PEP8NotebookBear(LocalBear):
+    LANGUAGES = {"Python", "Python 2", "Python 3"}
+    REQUIREMENTS = {PipRequirement('autopep8', '1.*'),
+                    PipRequirement('nbformat', '4.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/83333'
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: I just watched this - it's super awesome!

File: bears/python/PEP8NotebookBear.py [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+                   "select": pep_select,
+                   "max_line_length": max_line_length,
+                   "indent_size": indent_size}
+        notebook_node = notebook_node_from_list(file)
+        cells = notebook_node['cells']
+
+        for i, cell in enumerate(cells):
+            if cell['cell_type'] != 'code':
+                continue
+            source = cell['source']
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: ths doesn't get reused, can't you pass it directly?

File: bears/python/PEP8NotebookBear.py [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+            cells[i]['source'] = autopep8_fix_code_cell(source,
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: can't you do `cell['source']` instead and then you don't need enumerate

File: tests/python/PEP8NotebookBearTest.py [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.5.2"
```
Comment (sils1297) [inline tests/python/PEP8NotebookBearTest.py]: can you remove unrelated stuff so it get's shorter and only holds the relevant parts? Maybe one or two unrlated cells to show that they're not affected

File: bears/python/PEP8NotebookBear.py (hunk 18 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: ok, thanks for pointing that out

File: bears/python/PEP8NotebookBear.py (hunk 17 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: > How to name things: the hardest problem in programming

;) 
okay, let's call it a *string_list*

File: bears/python/PEP8NotebookBear.py (hunk 21 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: of course!
pls do not ask why it was like this ^^

Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: heh, I've seen worse. Look what I found during another code review: https://www.reddit.com/r/badcode/comments/4ufhnz/thats_how_we_roll_at_coala/

File: tests/python/PEP8NotebookBearTest.py (hunk 22 above)
Comment (caichinger) [inline tests/python/PEP8NotebookBearTest.py]: I removed the content from the metadata field, it is now much shorter.
When you apply the bear to the example file you see that it will skip the markdown cell and the one where there is no error. However, you do not see in in the test if this is what you meant?

File: bears/python/PEP8NotebookBear.py (hunk 21 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: very comforting ;)

**Comment by AbdealiJK** [comment]
@caichinger This is some awesome work. Some usability issues I can think of:

 - Do we check anywhere that the notebook/cell is actually a python notebook ? It can also be a R, Julia, etc notebook correct ?
 - Can't users use the magic keyworks like %pylab, %sql, etc. to actually inject non-python code into python notebooks ? What do we do about such things ?

I think these are something to look at later, so possibly create issues for them (doesn't need to be fixed in this PR)

**Comment by sils1297** [comment]
> Do we check anywhere that the notebook/cell is actually a python notebook ? It can also be a R, Julia, etc notebook correct ?

If the user feeds it to this bear it's his responsibility, in general our bears don't check the language, the user gives the language and the bear assumes that that's true. I think that's the right thing to do.

> Can't users use the magic keyworks like %pylab, %sql, etc. to actually inject non-python code into python notebooks ? What do we do about such things ?

Good question, IMO let's test it out once, see if it breaks everything - and it probably doesnt, worst case I think will be that that line might be treated wrongly - and just file an issue for everything that comes up.

**Comment by AbdealiJK** [comment]
While it's true that we let the use decide which bear to use when,
generally is easy because of the file extension. Here, all of them would
have .ipynb and the user would need to specify each file individually.
Hence a usability issue

On Sunday, August 28, 2016, Lasse Schuirmann <notifications@github.com>
wrote:
> Do we check anywhere that the notebook/cell is actually a python notebook
? It can also be a R, Julia, etc notebook correct ?
>
> If the user feeds it to this bear it's his responsibility, in general our
bears don't check the language, the user gives the language and the bear
assumes that that's true. I think that's the right thing to do.
>
> Can't users use the magic keyworks like %pylab, %sql, etc. to actually
inject non-python code into python notebooks ? What do we do about such
things ?
>
> Good question, IMO let's test it out once, see if it breaks everything -
and it probably doesnt, worst case I think will be that that line might be
treated wrongly - and just file an issue for everything that comes up.

**Comment by sils1297** [comment]
are the other languaged notebooks really called I*python*notebooks?

**Comment by sils1297** [comment]
does the notebook accidentally have the language embedded for each cell or something?

**Comment by mr-karan** [comment]
> https://github.com/coala-analyzer/coala-bears/pull/655#issuecomment-242964926

No iJulia, iRKernel etc

**Comment by AbdealiJK** [comment]
No idea what they're called but the extension is .ipynb
I think there all just called 'notebooks' with different kernels. I can dynamically change the kernel for a notebook

**Comment by caichinger** [comment]
> @caichinger This is some awesome work.

Thank you for your support :)

There are indeed a few things that should be kept in mind, however, they are imo out of the scope of this bear.

Sooner or later I would like to move this (necessary) discussion to some other place because it is more general in nature. There we could set up/maintain a document (place for discussion) to specify requirements and collect possible issues we want to keep in mind. @sils1297 @AbdealiJK : what do you think? I can prepare a draft it you tell me where to put it. Is Jupyter Notebook stuff a case for a coala enhancement proposal or is this topic too specific?

---

Regarding magics (line magic % (only affects the line), cell magic %% (affects the entire cell)):
- Since the user is asked whether a change should be applied or not, I think for now it is okay to leave it like it is.
- However, to provide a better user experience we should take care of them in the long run. For example you can insert %%javascript or %%cython cells which should then be skipped automatically. Note that these magics are part of the IPython kernel and not of Juypter (maybe there are similar features for other kernels but I do not know about them).

---

Regarding programming language:
- I also think it is okay to assume that if the user applies a PEP8 bear then it is about Python. However, for cross-language projects (e.g. I mix Python and R) it would be handy to say something like: apply all these checks to these notebooks and select the appropriate bear for each case/file (yet, this is not something the PEP8NotebookBear has do deal with).
- As already pointed out, the extension is *.ipynb*.
- The information about the used kernel is provided in the notebook's metadata field (which is empty an notebook creation but filled as soon as you put some content in). The matadata field looks like this:

**R**:
```
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
```

**Python**:
```
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
```
- As already mentioned, each cell has a cell type which can be e.g. 'code', 'markdown', ... There is no language information attached to each cell.

File: bears/python/PEP8NotebookBear.py [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    """
+    Writes a NotebookNode to a list of strings.
+
+    :param notebook_node: The notebook as NotebookNode to write.
+    :return:              The notebook as list of strings (linewise).
+    """
+    notebook_node_as_string_list = (nbformat.writes(notebook_node,
+                                                    nbformat.NO_CONVERT)
+                                    .splitlines(True))
+    return notebook_node_as_string_list
```
Comment (sils1297) [inline bears/python/PEP8NotebookBear.py]: just return it directly

**Comment by chauffer** [comment]
Hi, I'm sils trying out if our access policies are correct

**Comment by chauffer** [comment]
@rultor merge

**Comment by sils1297** [comment]
commit message Fixes -> Closes please

File: bears/python/PEP8NotebookBear.py (hunk 23 above)
Comment (caichinger) [inline bears/python/PEP8NotebookBear.py]: sure! :)

**Comment by sils1297** [comment]
Hey,

we use some bug prediction in coala which relies on the `Fixes` keyword in commit messages. To get good results from that we need to use `Closes` for normal issues instead of `Fixes` which should only be used for real bugs. (See also [the commit message docs](http://coala.readthedocs.io/en/latest/Getting_Involved/Writing_Good_Commits.html).)

To change your message you just use `git commit --amend` and then `git push --force` the new commit to replace the old one.

If you're just looking to fix an issue very quickly and not interested in contributing to coala long term, we can fix up the message for you - just tell us :).

Cheers!

File: tests/python/PEP8NotebookBearTest.py [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+    """
+    Utility function to prepare Notebook strings for further processing.
+
+    replace() necessary to avoid
+    nbformat.reader.NotJSONError: Notebook does not appear to be JSON
+    which is caused by
+    json.decoder.JSONDecodeError: Invalid control character
+    which is caused by the \n characters inside the input cells
+    """
+    return file.replace("\n\"", "\\n\"")
```
Comment (sils1297) [inline tests/python/PEP8NotebookBearTest.py]: instead of preparing it with that function you could just do a raw string, right? `r"\n" == "\\n"`

**Comment by sils1297** [comment]
Hi,

it looks like your PR is out of date and needs a rebase. [This page](http://coala.readthedocs.org/en/latest/Users/Tutorials/Git_Help.html) may help you to get started on this. (*Do not press the update branch button. It's evil.*)

Add the original coala repository as a remote:

```
git remote add upstream https://github.com/coala-analyzer/coala.git
```

Then go to your branch and rebase it onto our master:

```
git checkout <yourbranch>
git fetch upstream
git rebase --interactive upstream/master
```

Now push it up. You will need to do a force push because you change history here (do not do this on master branches :)):

```
git push --force
```

If you're just looking to fix an issue very quickly and not interested in contributing to coala long term, we can fix it up for you - just tell us :).

Ping us on [our gitter channel](https://gitter.im/coala-analyzer/coala) if you need help!

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/655/commits/a94f6dce53df982116457c47db480a52996b0efa

**Comment by sils1297** [comment]
but needs a rebase

File: tests/python/PEP8NotebookBearTest.py (hunk 24 above)
Comment (caichinger) [inline tests/python/PEP8NotebookBearTest.py]: Omg, I wrote a function for that, yes. At the time of writing I did not see through it...
Thanks! :+1:

**Comment by AbdealiJK** [comment]
Sorry, but you need a rebase again.
It can get merged as soon as the rebase is done though

**Comment by caichinger** [comment]
@AbdealiJK no problem, I have just rebased :)

**Comment by sils1297** [comment]
@caichinger let's see if everything passes now, then we shouldbe able to merge it right ahead

**Comment by sils1297** [comment]
I'll cherry pick this for your convenience :)

**Comment by sils1297** [comment]
cherry picked to https://github.com/coala-analyzer/coala-bears/pull/766
